// @nomicfoundation/ignition-core/dist/src/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyErrorMessageTemplate = exports.NomicIgnitionPluginError = exports.IgnitionPluginError = exports.IgnitionError = exports.CustomError = void 0;
const errors_list_1 = require("./errors-list");
/**
 * Base error class extended by all custom errors.
 * Placeholder to allow us to customize error output formatting in the future.
 *
 * @alpha
 */
class CustomError extends Error {
    constructor(message, cause) {
        super(message, cause !== undefined ? { cause } : undefined);
        this.name = this.constructor.name;
    }
}
exports.CustomError = CustomError;
/**
 * All exceptions intentionally thrown with Ignition-core
 * extend this class.
 *
 * @alpha
 */
class IgnitionError extends CustomError {
    // We store the error descriptor as private field to avoid
    // interferring with Node's default error formatting.
    // We can use getters to access any private field without
    // interferring with it.
    //
    // Disabling this rule as private fields don't use `private`
    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
    #errorDescriptor;
    constructor(errorDescriptor, messageArguments = {}, cause) {
        const prefix = `${(0, errors_list_1.getErrorCode)(errorDescriptor)}: `;
        const formattedMessage = applyErrorMessageTemplate(errorDescriptor.message, messageArguments);
        super(prefix + formattedMessage, cause);
        this.#errorDescriptor = errorDescriptor;
    }
    get errorNumber() {
        return this.#errorDescriptor.number;
    }
}
exports.IgnitionError = IgnitionError;
/**
 * This class is used to throw errors from Ignition plugins made by third parties.
 *
 * @alpha
 */
class IgnitionPluginError extends CustomError {
    static isIgnitionPluginError(error) {
        return (typeof error === "object" &&
            error !== null &&
            error._isIgnitionPluginError === true);
    }
    _isIgnitionPluginError = true;
    pluginName;
    constructor(pluginName, message, cause) {
        super(message, cause);
        this.pluginName = pluginName;
    }
}
exports.IgnitionPluginError = IgnitionPluginError;
/**
 * This class is used to throw errors from *core* Ignition plugins.
 * If you are developing a third-party plugin, use IgnitionPluginError instead.
 *
 * @alpha
 */
class NomicIgnitionPluginError extends IgnitionPluginError {
    static isNomicIgnitionPluginError(error) {
        return (typeof error === "object" &&
            error !== null &&
            error._isNomicIgnitionPluginError === true);
    }
    _isNomicIgnitionPluginError = true;
}
exports.NomicIgnitionPluginError = NomicIgnitionPluginError;
/**
 * This function applies error messages templates like this:
 *
 *  - Template is a string which contains a variable tags. A variable tag is a
 *    a variable name surrounded by %. Eg: %plugin1%
 *  - A variable name is a string of alphanumeric ascii characters.
 *  - Every variable tag is replaced by its value.
 *  - %% is replaced by %.
 *  - Values can't contain variable tags.
 *  - If a variable is not present in the template, but present in the values
 *    object, an error is thrown.
 *
 * @param template - The template string.
 * @param values - A map of variable names to their values.
 *
 * @alpha
 */
function applyErrorMessageTemplate(template, values) {
    return _applyErrorMessageTemplate(template, values, false);
}
exports.applyErrorMessageTemplate = applyErrorMessageTemplate;
function _applyErrorMessageTemplate(template, values, isRecursiveCall) {
    if (!isRecursiveCall) {
        for (const variableName of Object.keys(values)) {
            if (variableName.match(/^[a-zA-Z][a-zA-Z0-9]*$/) === null) {
                throw new IgnitionError(errors_list_1.ERRORS.INTERNAL.TEMPLATE_INVALID_VARIABLE_NAME, {
                    variable: variableName,
                });
            }
            const variableTag = `%${variableName}%`;
            if (!template.includes(variableTag)) {
                throw new IgnitionError(errors_list_1.ERRORS.INTERNAL.TEMPLATE_VARIABLE_NOT_FOUND, {
                    variable: variableName,
                });
            }
        }
    }
    if (template.includes("%%")) {
        return template
            .split("%%")
            .map((part) => _applyErrorMessageTemplate(part, values, true))
            .join("%");
    }
    for (const variableName of Object.keys(values)) {
        let value;
        if (values[variableName] === undefined) {
            value = "undefined";
        }
        else if (values[variableName] === null) {
            value = "null";
        }
        else {
            value = values[variableName].toString();
        }
        if (value === undefined) {
            value = "undefined";
        }
        const variableTag = `%${variableName}%`;
        if (value.match(/%([a-zA-Z][a-zA-Z0-9]*)?%/) !== null) {
            throw new IgnitionError(errors_list_1.ERRORS.INTERNAL.TEMPLATE_VALUE_CONTAINS_VARIABLE_TAG, { variable: variableName });
        }
        template = template.split(variableTag).join(value);
    }
    return template;
}
